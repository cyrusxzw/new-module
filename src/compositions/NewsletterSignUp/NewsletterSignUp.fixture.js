import React from 'react'; // imported because the html tag will be passed as props
const NewsletterSignUpFixture = {
  consentErrorMsg:
    'Please confirm your accept our Terms & Conditions before proceeding ',
  errorMessage: 'Please enter valid email address',
  termsAndCondition:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque quis sem quis sem bibendum aliquet non nec sapien. Suspendisse felis dui, gravida non ligula id, tempor pharetra sapien. Vestibulum porttitor felis vitae mollis porta. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque quis sem quis sem bibendum aliquet non nec sapien. Suspendisse felis dui, gravida non ligula id, tempor pharetra sapien. Vestibulum porttitor felis vitae mollis porta. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque quis sem quis sem bibendum aliquet non nec sapien. Suspendisse felis dui, gravida non ligula id, tempor pharetra sapien. Vestibulum porttitor felis vitae mollis porta. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque quis sem quis sem bibendum aliquet non nec sapien. Suspendisse felis dui, gravida non ligula id, tempor pharetra sapien. Vestibulum porttitor felis vitae mollis porta.',
  termsMessage: 'I accept the Terms & Conditions ',
  subscriptionMessage: (
    <span>
      By subscribing, you agree to receive email marketing from Aesop about our
      products and services, and accept our{' '}
      <a href="https://www.aesop.com/au/r/privacy/">privacy policy</a>.
    </span>
  ),
};

export { NewsletterSignUpFixture };
