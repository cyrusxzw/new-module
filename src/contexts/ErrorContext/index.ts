export { ErrorContextProvider, useErrorContext } from './ErrorContext';
