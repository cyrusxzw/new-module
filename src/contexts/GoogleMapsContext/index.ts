export {
  GoogleMapsContextProvider,
  useGoogleMapsContext,
} from './GoogleMapsContext';
