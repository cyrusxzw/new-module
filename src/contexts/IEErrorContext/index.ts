export { IEErrorContextProvider, useIEErrorContext } from './IEErrorContext';
