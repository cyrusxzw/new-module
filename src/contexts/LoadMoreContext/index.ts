export { LoadMoreContextProvider, useLoadMoreContext } from './LoadMoreContext';
