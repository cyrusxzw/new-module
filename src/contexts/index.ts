export {
  AddToCartContextProvider,
  useAddToCartContext,
} from './AddToCartContext';

export {
  GoogleMapsContextProvider,
  useGoogleMapsContext,
} from './GoogleMapsContext';

export { ErrorContextProvider, useErrorContext } from './ErrorContext';

export { LoadMoreContextProvider, useLoadMoreContext } from './LoadMoreContext';

export {
  NavBarThemeContextProvider,
  useNavBarThemeContext,
} from './NavBarThemeContext';

export {
  NotificationContextProvider,
  useNotificationContext,
} from './NotificationContext';

export {
  ProductDetailContextProvider,
  useProductDetailContext,
} from './ProductDetailContext';

export { ThemeContextProvider, useThemeContext } from './ThemeContext';

export {
  VariantSelectContextProvider,
  useVariantSelectContext,
} from './VariantSelectContext';
