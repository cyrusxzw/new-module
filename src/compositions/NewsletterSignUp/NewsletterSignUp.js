import React, { useState } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { useNotificationContext } from '~/contexts';
import { Icon } from '~/components/Icon';
import { Loading } from '~/components/Loading';
import { Checkbox } from '~/components/Checkbox';
import { TextInput } from '~/components/TextInput/index.ts';
import styles from './NewsletterSignUp.css';

const NewsletterSignUp = ({
  className,
  consentErrorMsg,
  errorMessage,
  onClick,
  showTermsConditionsTextBox,
  termsAndCondition,
  termsMessage,
  theme,
  subscriptionMessage,
}) => {
  const [hasContent, setHasContent] = useState(false);
  const [email, setEmail] = useState(undefined);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(undefined);
  const [hasAcceptedSubscription, setHasAcceptedSubscription] = useState(false);
  const [hasAcceptedTerms, setHasAcceptedTerms] = useState(false);

  const inputFieldClassSet = cx(styles.inputField, styles[theme], className);
  const buttonClassSet = cx(
    styles.button,
    styles[theme],
    !!error && styles.hasError,
  );
  const signUpIconClassSet = cx(styles.signUpIcon, styles[theme]);
  const loadingClassSet = cx(styles.loading, styles[theme]);
  const subscriptionMessageClassSet = cx(
    styles.subscriptionMessage,
    styles[theme],
  );

  const footerSuccessModal = useNotificationContext();
  const { actionType, dispatch } = footerSuccessModal;

  const validate = (value) => {
    const EMAIL_REGEXP = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i;
    const string = String(value);
    if (!string) {
      return true;
    }

    const isValid = !!string.match(EMAIL_REGEXP);
    return isValid;
  };

  const onChange = (e) => {
    setEmail(e.target.value);
    setHasContent(e.target.value && e.target.value.trim() !== '');
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    setError(undefined);
    const isValid = validate(email);
    if (
      isValid &&
      !(
        showTermsConditionsTextBox &&
        (!hasAcceptedSubscription || !hasAcceptedTerms)
      )
    ) {
      setIsLoading(true);
      await onClick(email);
      setIsLoading(false);
      dispatch({ type: actionType });
    } else {
      let errorMsg = errorMessage;
      if (
        showTermsConditionsTextBox &&
        (!hasAcceptedSubscription || !hasAcceptedTerms)
      ) {
        errorMsg = consentErrorMsg;
      }
      setError(errorMsg);
    }
  };

  return (
    <div className={styles.base}>
      <form onSubmit={handleSubmit}>
        <div className={inputFieldClassSet}>
          <div className={styles.footerNewsletterInputWrapper}>
            <TextInput
              data-test-ref="EMAIL_SIGN_UP_INPUT"
              hasContent={hasContent}
              hasError={!!error}
              id="newsletter-email"
              inputClassName={cx(styles.footerNewsletter)}
              isEnabled={!isLoading}
              label="Email Address"
              name="newsletter-email"
              onChange={onChange}
              theme={theme}
            />
          </div>
          <button
            className={buttonClassSet}
            data-test-ref="EMAIL_SIGN_UP_BUTTON"
            disabled={!!isLoading || !hasContent}
          >
            <Icon
              className={signUpIconClassSet}
              name="rightArrow"
              theme={theme}
            />
          </button>
          {!!isLoading && (
            <Loading
              className={loadingClassSet}
              isLoading={true}
              theme={theme}
            />
          )}
        </div>
        <div className={styles.footerNewsletterErrorMessage}>{error}</div>
        {showTermsConditionsTextBox ? (
          <section>
            <Checkbox
              content={subscriptionMessage}
              dataTestRef="EMAIL_SIGN_UP_SUBSCRIPTION_CHECKBOX"
              id="subscription"
              isEnabled={!isLoading}
              onChange={() =>
                setHasAcceptedSubscription(!hasAcceptedSubscription)
              }
              theme={theme}
            />
            <Checkbox
              content={termsMessage}
              dataTestRef="EMAIL_SIGN_UP_TERMS_AND_CONDITIONS_CHECKBOX"
              id="terms"
              isEnabled={!isLoading}
              onChange={() => setHasAcceptedTerms(!hasAcceptedTerms)}
              theme={theme}
            />
            <div className={styles.termsAndConditionWrapper}>
              <div className={styles.termsAndCondition}>
                {termsAndCondition}
              </div>
            </div>
          </section>
        ) : (
          <div className={subscriptionMessageClassSet}>
            {subscriptionMessage}
          </div>
        )}
      </form>
    </div>
  );
};

NewsletterSignUp.propTypes = {
  className: PropTypes.string,
  consentErrorMsg: PropTypes.string,
  errorMessage: PropTypes.string,
  onClick: PropTypes.func,
  showTermsConditionsTextBox: PropTypes.bool,
  termsAndCondition: PropTypes.string,
  termsMessage: PropTypes.string,
  theme: PropTypes.string,
  subscriptionMessage: PropTypes.string,
};

NewsletterSignUp.defaultProps = {
  className: undefined,
  consentErrorMsg: undefined,
  errorMessage: 'NEWSLETTER_SIGNUP',
  onClick: undefined,
  showTermsConditionsTextBox: false,
  termsAndCondition: undefined,
  termsMessage: undefined,
  theme: 'dark',
  subscriptionMessage: undefined,
};

export { NewsletterSignUp };
