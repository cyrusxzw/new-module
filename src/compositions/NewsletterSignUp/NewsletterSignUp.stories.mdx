import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { NotificationContextProvider } from '~/contexts';
import { Podium } from '~/components/Podium';
import { NotificationModal } from '~/components/NotificationModal';
import { NotificationModalFixture } from '~/components/NotificationModal/NotificationModal.fixture';
import { NewsletterSignUp } from './NewsletterSignUp';
import { NewsletterSignUpFixture } from './NewsletterSignUp.fixture';
import { newsletterSignUpOnClick } from './__mocks__/NewsletterSignUp.onClick';

# NewsletterSignUp

<Meta title="Compositions/NewsletterSignUp" component={NewsletterSignUp} />

<Canvas>
  <Story name="Base component">
    <Podium horizontalPadding="small" verticalPadding="small">
      <NotificationContextProvider>
        <div style={{ position: 'relative' }}>
          <NotificationModal
            backgroundColor="#d5d5cc"
            notificationMessage={NotificationModalFixture.notificationMessage}
          />
          <NewsletterSignUp
            consentErrorMsg={NewsletterSignUpFixture.consentErrorMsg}
            errorMessage={NewsletterSignUpFixture.errorMessage}
            onClick={newsletterSignUpOnClick}
            showTermsConditionsTextBox={knobs.boolean(
              'showTermsConditionsTextBox',
              true,
              'Presentation',
            )}
            subscriptionMessage={NewsletterSignUpFixture.subscriptionMessage}
            termsAndCondition={NewsletterSignUpFixture.termsAndCondition}
            termsMessage={NewsletterSignUpFixture.termsMessage}
            theme={knobs.select(
              'theme',
              ['dark', 'light'],
              'dark',
              'Presentation',
            )}
          />
        </div>
      </NotificationContextProvider>
    </Podium>
  </Story>
</Canvas>

## API

<ArgsTable of={NewsletterSignUp} />
