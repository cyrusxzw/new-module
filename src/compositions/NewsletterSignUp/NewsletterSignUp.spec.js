import React from 'react';
import { render } from '@testing-library/react';
import { NotificationContextProvider } from '~/contexts';
import { NewsletterSignUp } from './NewsletterSignUp';
import { NewsletterSignUpFixture } from './NewsletterSignUp.fixture';

describe('<NewsletterSignUp />', () => {
  it('should render base component correctly', () => {
    const { container } = render(
      <NotificationContextProvider>
        <NewsletterSignUp
          consentErrorMsg={NewsletterSignUpFixture.consentErrorMsg}
          errorMessage={NewsletterSignUpFixture.errorMessage}
          showTermsConditionsTextBox={true}
          subscriptionMessage={NewsletterSignUpFixture.subscriptionMessage}
          termsAndCondition={NewsletterSignUpFixture.termsAndCondition}
          termsMessage={NewsletterSignUpFixture.termsMessage}
          theme="light"
        />
      </NotificationContextProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
