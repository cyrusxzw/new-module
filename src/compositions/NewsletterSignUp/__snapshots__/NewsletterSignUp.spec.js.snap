// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewsletterSignUp /> should render base component correctly 1`] = `
<div>
  <div
    class="base"
  >
    <form>
      <div
        class="inputField light"
      >
        <div
          class="footerNewsletterInputWrapper"
        >
          <div
            class="formText"
          >
            <label
              for="newsletter-email"
            >
              <input
                class="formTextInput light footerNewsletter"
                data-test-ref="EMAIL_SIGN_UP_INPUT"
                id="newsletter-email"
                name="newsletter-email"
                novalidate=""
                type="text"
                value=""
              />
              <span
                class="formTextLabel light"
              >
                Email Address
              </span>
            </label>
          </div>
        </div>
        <button
          class="button light"
          data-test-ref="EMAIL_SIGN_UP_BUTTON"
          disabled=""
        >
          <svg
            aria-labelledby="rightArrow-1"
            class="base rightArrow light signUpIcon light"
            data-testid="data-testid-Icon"
            focusable="false"
            height="12"
            role="img"
            style="height: 12px; width: 12px;"
            tabindex="-1"
            viewBox="0 0 50 50"
            width="12"
          >
            <title
              id="rightArrow-1"
            >
              rightArrow
            </title>
            <g>
              <path
                d="M30.1,5.3L50,25.1L30.1,45h-6.6l18-17.6H0v-4.8h41.5l-18-17.6h6.6V5.3z"
              />
            </g>
          </svg>
        </button>
      </div>
      <div
        class="footerNewsletterErrorMessage"
      />
      <section>
        <div>
          <label
            class="base"
            for="subscription"
          >
            <input
              aria-describedby="subscription-error-message"
              aria-invalid="false"
              class="checkBox light"
              data-test-ref="EMAIL_SIGN_UP_SUBSCRIPTION_CHECKBOX"
              id="subscription"
              type="checkbox"
            />
            <span
              class="content light"
            >
              <span>
                By subscribing, you agree to receive email marketing from Aesop about our products and services, and accept our
                 
                <a
                  href="https://www.aesop.com/au/r/privacy/"
                >
                  privacy policy
                </a>
                .
              </span>
            </span>
          </label>
        </div>
        <div>
          <label
            class="base"
            for="terms"
          >
            <input
              aria-describedby="terms-error-message"
              aria-invalid="false"
              class="checkBox light"
              data-test-ref="EMAIL_SIGN_UP_TERMS_AND_CONDITIONS_CHECKBOX"
              id="terms"
              type="checkbox"
            />
            <span
              class="content light"
            >
              I accept the Terms & Conditions 
            </span>
          </label>
        </div>
        <div
          class="termsAndConditionWrapper"
        >
          <div
            class="termsAndCondition"
          >
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque quis sem quis sem bibendum aliquet non nec sapien. Suspendisse felis dui, gravida non ligula id, tempor pharetra sapien. Vestibulum porttitor felis vitae mollis porta. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque quis sem quis sem bibendum aliquet non nec sapien. Suspendisse felis dui, gravida non ligula id, tempor pharetra sapien. Vestibulum porttitor felis vitae mollis porta. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque quis sem quis sem bibendum aliquet non nec sapien. Suspendisse felis dui, gravida non ligula id, tempor pharetra sapien. Vestibulum porttitor felis vitae mollis porta. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque quis sem quis sem bibendum aliquet non nec sapien. Suspendisse felis dui, gravida non ligula id, tempor pharetra sapien. Vestibulum porttitor felis vitae mollis porta.
          </div>
        </div>
      </section>
    </form>
  </div>
</div>
`;
